### ISSUE TITLE
Fix creation of directories during recovery for empty destination scenarios

### ISSUE THREAD

**Issue by mikewallace1979** [EnterpriseDB/barman#465]
The creation of destination directories during recovery is handled by [RsyncCopyController.create_dir_and_purge](https://github.com/EnterpriseDB/barman/blob/749f33bcfdd6ea865390ddb34ece95301c797690/barman/copy_controller.py#L957-L978) which relies on a filter (a file referenced by `exclude_and_protect_file`) to ensure that:

1. The source files are not copied.
2. Any existing files at the destination which are present on the source are not deleted.

This can take in the order of hours when large numbers of files are involved (see #377) therefore an optimisation was made in #378 where we avoid adding files to `exclude_and_protect` file if we know the destination is empty.

Unfortunately this created a new problem in that the rsync command in [RsyncCopyController.create_dir_and_purge](https://github.com/EnterpriseDB/barman/blob/749f33bcfdd6ea865390ddb34ece95301c797690/barman/copy_controller.py#L957-L978) no longer excludes the backup files. It therefore copies the backup files into the destination during the create directories step. *This has a severe performance cost* because the entire directory is copied in a single rsync process, rather than using the number of workers specified in the `--jobs` flag. For a suitably large database this can increase restore times by an order of magnitude or more.

There is a second problem with the optimisation in #378 where we determine if the PGDATA destination is empty or not: Because the [pg_tblspc directory is created](https://github.com/EnterpriseDB/barman/blob/749f33bcfdd6ea865390ddb34ece95301c797690/barman/recovery_executor.py#L625-L679) before the [check for directory content](https://github.com/EnterpriseDB/barman/blob/749f33bcfdd6ea865390ddb34ece95301c797690/barman/copy_controller.py#L875-L877) we always determine the destination PGDATA to have content and therefore do not apply the optimisation.

We therefore need to fix this in two places:
1. Exclude `pg_tblspc` from the directory content check in [RsyncCopyController._analyze_directory](https://github.com/EnterpriseDB/barman/blob/749f33bcfdd6ea865390ddb34ece95301c797690/barman/copy_controller.py#L875-L877) so that if it is the only entry we still consider the destination to have no content.
2. For the case where there is no content in the destination directory: use wildcard filters to include all directories and exclude all files. This will ensure the directories are still created on the destination while the `exclude_and_protect` filter excludes the actual files so that they can be copied in parallel during the next stage of the recover operation.

**Comment by mikewallace1979** [comment]
# Notes from testing the proposed fix #466

Tested manually by pausing Barman immediately after `_analyze_directory` to inspect the `exclude_and_protect_filter` file and immediately after the `_create_dir_and_purge` step to inspect the state of the destination directories.

Full recover command:

    barman recover --tablespace tbs1:/tmp/test_tbs1 --tablespace tbs2:/tmp/test_tbs2 --tablespace tbs3:/tmp/test_tbs3 test1 20211129T141025 /tmp/test_restore

Only `tbs1` has any content. After each test the content of the target directories is deleted.

## With release/2.13 (to confirm the original behaviour)

This is the <= 2.13 behaviour.
In the following output we can see the populated `exclude_and_protect_filter` file for the tablespace dir and for the pgdata dir.
We can also see that the `_create_dir_and_purge` function only creates the directories and does not copy the files over.
Looking at the barman code we can see the next step would be the [parallel execution of the copy commands](https://github.com/EnterpriseDB/barman/blob/release/2.13/barman/copy_controller.py#L556-L572).

### `tbs1`

#### After `_analyze_directory`

```
# cat /tmp/barman-eQdfDE/tbs1_exclude_and_protect.filter
P /PG_14_202107181/13697/16388
- /PG_14_202107181/13697/16388
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_tbs1/
/tmp/test_tbs1/
└── PG_14_202107181
    └── 13697

2 directories, 0 files
```

### `pgdata`

#### After `_analyze_directory`

```
# cat /tmp/barman-eQdfDE/pgdata_exclude_and_protect.filter
P /PG_VERSION
- /PG_VERSION
...
P /pg_xact/0000
- /pg_xact/0000
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_restore
/tmp/test_restore
├── base
│   ├── 1
│   ├── 13696
│   └── 13697
├── global
├── pg_commit_ts
├── pg_dynshmem
├── pg_logical
│   ├── mappings
│   └── snapshots
├── pg_multixact
│   ├── members
│   └── offsets
├── pg_notify
├── pg_replslot
├── pg_serial
├── pg_snapshots
├── pg_stat
├── pg_stat_tmp
├── pg_subtrans
├── pg_tblspc
│   ├── 16385 -> /tmp/test_tbs1
│   ├── 16386 -> /tmp/test_tbs2
│   └── 16387 -> /tmp/test_tbs3
├── pg_twophase
├── pg_wal
└── pg_xact

27 directories, 0 files
```

## With release/2.14 (to confirm the buggy behaviour)

Here we see the effect of the both bugs in the empty destination optimisation:
* Firstly, `tbs1` now has a file in the destination directory even though the main copy has not run - this tells us files were copied in the `_create_dir_and_purge` step.
* Secondly, the `exclude_and_protect.filter` file is not empty for the `pgdata` directory even though it now should be. As a side effect the files were not copied in the `_create_dir_and_purge` step.

### `tbs1`

#### After `_analyze_directory`

```
# cat /tmp/barman-v6rIz9/tbs1_exclude_and_protect.filter
#
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_tbs1
/tmp/test_tbs1
└── PG_14_202107181
    └── 13697
        └── 16388

2 directories, 1 file
```

### `pgdata`

#### After `_analyze_directory`

```
# cat /tmp/barman-v6rIz9/pgdata_exclude_and_protect.filter
P /PG_VERSION
- /PG_VERSION
...
P /pg_xact/0000
- /pg_xact/0000
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_restore/
/tmp/test_restore/
├── base
│   ├── 1
│   ├── 13696
│   └── 13697
├── global
├── pg_commit_ts
├── pg_dynshmem
├── pg_logical
│   ├── mappings
│   └── snapshots
├── pg_multixact
│   ├── members
│   └── offsets
├── pg_notify
├── pg_replslot
├── pg_serial
├── pg_snapshots
├── pg_stat
├── pg_stat_tmp
├── pg_subtrans
├── pg_tblspc
│   ├── 16385 -> /tmp/test_tbs1
│   ├── 16386 -> /tmp/test_tbs2
│   └── 16387 -> /tmp/test_tbs3
├── pg_twophase
├── pg_wal
└── pg_xact

27 directories, 0 files
```

## With PR #466 

Here we see everything working as intended.
No files are copied across for tbs1 during the `_create_dir_and_purge` step and its `exclude_and_protect.filter` file contains only the include directories and exclude files directives. This confirms the following:
* The files were not copied across by the single-worker `_create_dir_and_purge` rsync command, therefore the performance penalty of copying all the files via a single process does not occur.
* The intended outcome of the original optimisation should now be achieved, since the `exclude_and_protect.filter` file only has two entries instead of an entry for every item in the source.

Looking at the `pgdata` dir shows that again the `exclude_and_protect.filter` file only has the two entries and still no files are copied across.

### `tbs1`

#### After `_analyze_directory`

```
# cat /tmp/barman-2vgIMr/tbs1_exclude_and_protect.filter
+ */
- *
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_tbs1
/tmp/test_tbs1
└── PG_14_202107181
    └── 13697

2 directories, 0 files
```

### `pgdata`

#### After `_analyze_directory`

```
# cat /tmp/barman-2vgIMr/pgdata_exclude_and_protect.filter
+ */
- *
```

#### After `_create_dir_and_purge`

```
# tree /tmp/test_restore/
/tmp/test_restore/
├── base
│   ├── 1
│   ├── 13696
│   └── 13697
├── global
├── pg_commit_ts
├── pg_dynshmem
├── pg_logical
│   ├── mappings
│   └── snapshots
├── pg_multixact
│   ├── members
│   └── offsets
├── pg_notify
├── pg_replslot
├── pg_serial
├── pg_snapshots
├── pg_stat
├── pg_stat_tmp
├── pg_subtrans
├── pg_tblspc
│   ├── 16385 -> /tmp/test_tbs1
│   ├── 16386 -> /tmp/test_tbs2
│   └── 16387 -> /tmp/test_tbs3
├── pg_twophase
├── pg_wal
└── pg_xact

27 directories, 0 files
```

## With PR #466 when the destination directories are not empty

We have confirmed both bugs introduced in #378 and verified the fix in PR #466 fixes them both.

As a final check we can repeat the restore without deleting the target directories - we should see the `exclude_and_protect.filter` files are now fully populated with all source files and we should see the previously restored files still present after `_create_dir_and_purge`:

```
# cat /tmp/barman-pQGKlV/tbs1_exclude_and_protect.filter
P /PG_14_202107181/13697/16388
- /PG_14_202107181/13697/16388
```

```
# tree /tmp/test_tbs1
/tmp/test_tbs1
└── PG_14_202107181
    └── 13697
        └── 16388

2 directories, 1 file
```

```
# cat /tmp/barman-pQGKlV/pgdata_exclude_and_protect.filter
P /PG_VERSION
- /PG_VERSION
...
P /pg_xact/0000
- /pg_xact/0000
```

```
# tree /tmp/test_restore/
/tmp/test_restore/
├── backup_label
├── base
│   ├── 1
...
├── postgresql.auto.conf
└── postgresql.conf

27 directories, 949 files
``` 

A final check of the `pgdata` dir after the restore is fully complete:

```
# tree /tmp/test_restore
...
28 directories, 955 files
```

We can check the diff and see the difference in file count is due to the additional files created by `barman recover` being purged during the `_create_dir_and_purge` step:

```
# diff /tmp/before_full_restore <(tree /tmp/test_restore)
1c1
< /tmp/test_restore/
---
> /tmp/test_restore
973a974,978
> │   ├── 000000010000000000000005
> │   ├── 000000010000000000000005.00000028.backup
> │   └── archive_status
> │       ├── 000000010000000000000005.00000028.backup.done
> │       └── 000000010000000000000005.done
977c982,984
< └── postgresql.conf
---
> ├── postgresql.auto.conf.origin
> ├── postgresql.conf
> └── postgresql.conf.origin
979c986
< 27 directories, 949 files
---
> 28 directories, 955 files
```

We can see the missing files were the WALs, the `archive_status` dir and the `*.confi.origin` files and these are exactly the files that should have been purged. This confirms we haven't broken the existing behaviour.

**Comment by mikewallace1979** [comment]
Fixed in #466
